src/test/java/formflow/library/controllers/FileControllerTest.java

```diff
@@ -180,4 +190,113 @@ public void endpointDeletesFromCloudRepositoryDbAndSession() throws Exception {
+          .andReturn().getResponse().getContentAsByteArray();
+
+      assertThat(Arrays.equals(testcloudFile.getFileBytes(), response)).isTrue();
+    }
+
+    @Test
+    void multiFileEndpointShouldReturnZipOfUserFilesReturnedByTheCloudFileRepository() throws Exception {
+      session.setAttribute("id", submission.getId());
+      byte[] firstTestFileBytes = Files.readAllBytes(Paths.get("src/test/resources/test.png"));
+      byte[] secondTestFileBytes = Files.readAllBytes(Paths.get("src/test/resources/test-platypus.gif"));
+      long firstTestFileSize = firstTestFileBytes.length;
+      long secondTestFileSize = secondTestFileBytes.length;
+      CloudFile firstTestcloudFile = new CloudFile(firstTestFileSize, firstTestFileBytes);
+      CloudFile secondTestcloudFile = new CloudFile(secondTestFileSize, secondTestFileBytes);
+
+      UserFile firstTestUserFile = UserFile.builder().originalName("test.png").mimeType("image/png")
+          .repositoryPath("testPath")
+          .filesize((float) firstTestFileSize)
+          .submissionId(submission).build();
+      UserFile secondTestUserFile = UserFile.builder().originalName("test-platypus.gif").mimeType("image/gif")
+          .repositoryPath("testPath2")
+          .filesize((float) secondTestFileSize)
+          .submissionId(submission).build();
+
+      List<UserFile> userFiles = Arrays.asList(firstTestUserFile, secondTestUserFile);
+      when(userFileRepositoryService.findAllBySubmissionId(submission)).thenReturn(userFiles);
+      when(submissionRepositoryService.findById(submission.getId())).thenReturn(Optional.ofNullable(submission));
+      when(cloudFileRepository.get("testPath")).thenReturn(firstTestcloudFile);
+      when(cloudFileRepository.get("testPath2")).thenReturn(secondTestcloudFile);
+
+      MvcResult mvcResult = mockMvc.perform(
+              MockMvcRequestBuilders.get("/file-download/{submissionId}", submission.getId().toString())
+                  .session(session))
+          .andExpect(MockMvcResultMatchers.request().asyncStarted())
+          .andReturn();
+      byte[] response = mockMvc.perform(MockMvcRequestBuilders.asyncDispatch(mvcResult))
+          .andExpect(status().isOk())
+          .andExpect(header().string(HttpHeaders.CONTENT_DISPOSITION,
+              "attachment; filename=\"UserFiles-" + submission.getId() + ".zip\""))
+          .andReturn().getResponse().getContentAsByteArray();
+
+      byte[] testZipFile = Files.readAllBytes(Paths.get("src/test/resources/test-archive.zip"));
+
+      Map<String, byte[]> responseContents = getZipContentsMap(response);
+      Map<String, byte[]> testZipContents = getZipContentsMap(testZipFile);
+
+      assertThat(responseContents.keySet()).isEqualTo(testZipContents.keySet());
+
+      for (String name : responseContents.keySet()) {
+        assertThat(Arrays.equals(responseContents.get(name), testZipContents.get(name))).isTrue();
```

Comment:
We tried to do this but found that the sizes aren't consistent between MockMVC's return and the original file. I think it's because the response includes the bytes from the response headers which add more bytes.